File: pkg/gql/Slot.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2021 all rights reserved


# externals
import graphene

# my interface
from .Node import Node
# field types
from .Position import Position


# a factory slot
class Slot(graphene.ObjectType):
    """
    {Slot} captures the relationship between a {product} and a named slot in a {factory}
    """

    # {graphene} metadata
    class Meta:
        # register my interface
        interfaces = Node,

    # the fields
    id = graphene.ID()
    bound = graphene.Boolean(required=True)
    position = graphene.Field(Position, required=True)


    # the resolvers
    def resolve_id(slot, *_):
        return slot.relay

    def resolve_bound(slot, *_):
        return slot.product is not None

    def resolve_position(slot, *_):
        # unpack
        x, y = slot.position
        # build a position and return it
        return Position(x=x, y=y)


# end of file

File: ux/client/flogram/slot.js
// -*- web -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2021 all rights reserved


// externals
import React from 'react'
import { graphql, useFragment } from 'react-relay/hooks'
// project
import { Slot as Unbound } from '~/shapes'
import { Product as Bound } from '~/shapes'
import { useSelection } from '~/views/flo2d'
// local
import { Node } from './node'


// render the flow nodes that are bound/unbound slots
export const Slot = (props) => {
    // unpack the slot info
    const slot = useFragment(graphql`
        fragment slot_slot on Slot {
            id
            bound
            position {
                x
                y
            }
        }
    `, props.slot)

    // get the current selection
    const { selection } = useSelection()
    // am i selected
    const selected = selection.includes(slot.id)

    // render
    return (
        <Node id={slot.id} position={slot.position} >
            {slot.bound ? <Bound highlight={selected} /> : <Unbound highlight={selected} />}
        </Node>
    )
}


// end of file